tfhe/src/core_crypto/experimental/entities/lwe_shrinking_keyswitch_key.rs

```diff
@@ -0,0 +1,328 @@
+    ///     lwe_ksk.output_lwe_size(),
+    ///     output_lwe_dimension.to_lwe_size()
+    /// );
+    /// assert_eq!(
+    ///     lwe_ksk.shared_randomness(),
+    ///     lwe_secret_key_shared_coef_count
+    /// );
+    /// assert_eq!(
+    ///     lwe_ksk.unshared_randomness(),
+    ///     LweSecretKeyUnsharedCoefCount(input_lwe_dimension.0 - lwe_secret_key_shared_coef_count.0)
+    /// );
+    /// assert_eq!(lwe_ksk.ciphertext_modulus(), ciphertext_modulus);
+    ///
+    /// // Demonstrate how to recover the allocated container
+    /// let underlying_container: Vec<u64> = lwe_ksk.into_container();
+    ///
+    /// // Recreate a secret key using from_container
+    /// let lwe_ksk = LweShrinkingKeyswitchKey::from_container(
+    ///     underlying_container,
+    ///     decomp_base_log,
+    ///     decomp_level_count,
+    ///     input_lwe_dimension,
+    ///     output_lwe_dimension.to_lwe_size(),
+    ///     ciphertext_modulus,
+    /// );
+    ///
+    /// assert_eq!(lwe_ksk.decomposition_base_log(), decomp_base_log);
+    /// assert_eq!(lwe_ksk.decomposition_level_count(), decomp_level_count);
+    /// assert_eq!(lwe_ksk.input_key_lwe_dimension(), input_lwe_dimension);
+    /// assert_eq!(lwe_ksk.output_key_lwe_dimension(), output_lwe_dimension);
+    /// assert_eq!(
+    ///     lwe_ksk.output_lwe_size(),
+    ///     output_lwe_dimension.to_lwe_size()
+    /// );
+    /// assert_eq!(
+    ///     lwe_ksk.shared_randomness(),
+    ///     lwe_secret_key_shared_coef_count
+    /// );
+    /// assert_eq!(
+    ///     lwe_ksk.unshared_randomness(),
+    ///     LweSecretKeyUnsharedCoefCount(input_lwe_dimension.0 - lwe_secret_key_shared_coef_count.0)
+    /// );
+    /// assert_eq!(lwe_ksk.ciphertext_modulus(), ciphertext_modulus);
+    /// ```
+    pub fn from_container(
+        container: C,
+        decomp_base_log: DecompositionBaseLog,
+        decomp_level_count: DecompositionLevelCount,
+        input_key_lwe_dimension: LweDimension,
+        output_lwe_size: LweSize,
+        ciphertext_modulus: CiphertextModulus<C::Element>,
+    ) -> Self {
+        let lwe_ksk = LweKeyswitchKey::from_container(
+            container,
+            decomp_base_log,
+            decomp_level_count,
+            output_lwe_size,
+            ciphertext_modulus,
+        );
+
+        let output_key_lwe_dimension = lwe_ksk.output_key_lwe_dimension();
+        let unshared_randomness_coef_count =
+            LweSecretKeyUnsharedCoefCount(lwe_ksk.input_key_lwe_dimension().0);
+        let shared_randomness_coef_count =
+            input_key_lwe_dimension.shared_coef_count_from(unshared_randomness_coef_count);
+
+        assert!(
+            output_key_lwe_dimension.0 <= input_key_lwe_dimension.0,
+            "The output LweDimension ({output_key_lwe_dimension:?}) \
+                must be smaller than the input LweDimension ({input_key_lwe_dimension:?}) \
+                for an LweShrinkingKeyswitchKey."
+        );
```

Comment:
Is this necessary?

Reply:
maybe not 🤔

Reply:
same thing shrinking for shared keys means the output key needs to be smaller